src/mpid/ch4/shm/posix/eager/iqueue/iqueue_send.h

```diff
@@ -71,59 +71,44 @@ MPIDI_POSIX_eager_send(int grank,
 
     /* Figure out the capacity of each cell */
     capacity = MPIDI_POSIX_EAGER_IQUEUE_CELL_CAPACITY(transport);
+
     available = capacity;
 
     cell->from = MPIDI_POSIX_global.my_local_rank;
 
     /* If this is the beginning of the message, mark it as the head. Otherwise it will be the
      * tail. */
-    if (*msg_hdr) {
+    cell->payload_size = 0;
+    if (!offset) {
         cell->am_header = **msg_hdr;
-        *msg_hdr = NULL;        /* completed header copy */
         cell->type = MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_HDR;
+        /* send am_hdr if this is the first segment */
+        MPIR_Typerep_copy(payload, am_hdr, am_hdr_sz);
+        /* make sure the data region starts at the boundary of MAX_ALIGNMENT */
+        payload = payload + MPIDI_POSIX_RESIZE_TO_MAX_ALIGN(am_hdr_sz);
+        cell->payload_size += MPIDI_POSIX_RESIZE_TO_MAX_ALIGN(am_hdr_sz);
+        cell->am_header.am_hdr_sz = MPIDI_POSIX_RESIZE_TO_MAX_ALIGN(am_hdr_sz);
+        available -= cell->am_header.am_hdr_sz;
     } else {
         cell->type = MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_DATA;
     }
 
-    /* Pack the data into the cells */
-    iov_done = 0;
-    for (i = 0; i < *iov_num; i++) {
-        /* Optimize for the case where the message will fit into the cell. */
-        if (likely(available >= (*iov)[i].iov_len)) {
-            MPIR_Typerep_copy(payload, (*iov)[i].iov_base, (*iov)[i].iov_len);
-
-            payload += (*iov)[i].iov_len;
-            available -= (*iov)[i].iov_len;
-
-            iov_done++;
+    if (data_sz_left) {
```

Comment:
If `data_sz_left`, I think this function should not be called in the first place, or we could use `MPI_DATATYPE_NULL` (and `count=0`) to indicate there is no data.

Reply:
Sure.

Reply:
Actually, I take that back. I would prefer checking (data_sz_left == 0). Although it is equivalent to checking if `count == 0` or `datatype == MPI_DATATYPE_NULL`. But those two are valid MPI send parameters per standard. The check here is really to differentiate between the header-only and non-header-only case.

Reply:
The eager interface is not responsible for the MPI layer though, it is supposed to get processed input from posix layer. To clarify:

* case 1: header only -- it needs `am_hdr+am_hdr_sz`, and `buf==NULL`
* case 2: header + first segment -- it needs `am_hdr+am_hdr_sz` and `buf-count-datatype`, and `offset == 0`
* case 3: additional segments: `am_hdr==NULL` and `buf-count-datatype` plus `offset`

Did I miss anything?

Reply:
That is right. I did not mean it responsible for MPI layer. I think it would be better to avoid relying on something that may or may not run into problem with upper layer.